cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -490,6 +516,151 @@ private void closeScanner() {
     }
   }
 
+  /**
+   * Filters elements matching an index  {@link ScannerIterator}.
+   */
+  @VisibleForTesting
+  static final class FilterByIndexIterator extends AbstractCloseableIterator<StructuredRow> {
+    private final Iterator<? extends CloseableIterator<StructuredRow>> scannerIterator;
+    private CloseableIterator<StructuredRow> currentScanner;
+    private final Field<?> filterIndex;
+    private final StructuredTableSchema schema;
+
+    FilterByIndexIterator(Iterator<? extends CloseableIterator<StructuredRow>> scannerIterator, Field<?> filterIndex,
+                          StructuredTableSchema schema) {
+      this.scannerIterator = scannerIterator;
+      this.currentScanner = scannerIterator.hasNext() ? scannerIterator.next() : CloseableIterator.empty();
+      this.filterIndex = filterIndex;
+      this.schema = schema;
+    }
+
+    @Override
+    protected StructuredRow computeNext() {
+      // Postfiltering on scanned rows by index
+      while (!currentScanner.hasNext() && scannerIterator.hasNext()) {
+        closeScanner();
+        currentScanner = scannerIterator.next();
+      }
+      while (currentScanner.hasNext()) {
+        StructuredRow row = currentScanner.next();
+        if (Objects.equals(getFromRow(row), filterIndex.getValue())) {
+          return row;
+        }
+      }
+      return endOfData();
+    }
+
+    @Override
+    public void close() {
+      closeScanner();
+    }
+
+    private Object getFromRow(StructuredRow row) throws InvalidFieldException {
+      switch (filterIndex.getFieldType()) {
+        case INTEGER:
+          return row.getInteger(filterIndex.getName());
+        case LONG:
+          return row.getLong(filterIndex.getName());
+        case FLOAT:
+          return row.getFloat(filterIndex.getName());
+        case DOUBLE:
+          return row.getDouble(filterIndex.getName());
+        case STRING:
+          return row.getString(filterIndex.getName());
+        case BYTES:
+          return row.getBytes(filterIndex.getName());
+        default:
+          throw new InvalidFieldException(schema.getTableId(), filterIndex.getName());
+      }
+    }
+
+    private void closeScanner() {
+      if (currentScanner != null) {
+        currentScanner.close();
+        currentScanner = null;
+      }
+    }
+  }
+
+  /**
+   * Sorts elements based off a column  {@link ScannerIterator}.
+   */
+  @VisibleForTesting
+  static final class SortedIterator extends AbstractCloseableIterator<StructuredRow> {
+    private final Iterator<? extends CloseableIterator<StructuredRow>> scannerIterator;
```

Comment:
This one can be a local variable?

Reply:
I will remove this iterator as per previous comment, so won't need this